index.js

```diff
@@ -4,15 +4,125 @@
 
   /*istanbul ignore next*/
   if(module && typeof module.exports !== 'undefined'){
-    module.exports = f();
+    module.exports = f(require('sanctuary-type-classes'), require('sanctuary-type-identifiers'));
   }else{
-    global.concurrify = f();
+    global.concurrify = f(global.sanctuaryTypeClasses, global.sanctuaryTypeIdentifiers);
   }
 
-}(/*istanbul ignore next*/(global || window || this), function(){
+}(/*istanbul ignore next*/(global || window || this), function(Z, type){
 
   'use strict';
 
-  return function(){};
+  var $alt = 'fantasy-land/alt';
+  var $ap = 'fantasy-land/ap';
+  var $map = 'fantasy-land/map';
+  var $of = 'fantasy-land/of';
+  var $zero = 'fantasy-land/zero';
+  var $$type = '@@type';
+
+  var ordinal = ['first', 'second', 'third', 'fourth', 'fifth'];
+  var show = Z.toString;
+
+  function isFunction(f){
+    return typeof f === 'function';
+  }
+
+  function isBinary(f){
+    return f.length >= 2;
+  }
+
+  function isApplicativeRepr(Repr){
+    return isFunction(Repr[$of]) && Z.Applicative.test(Repr[$of]());
+  }
+
+  function invalidArgument(it, at, expected, actual){
+    throw new TypeError(
+      it
+      + ' expects its '
+      + ordinal[at]
+      + ' argument to '
+      + expected
+      + '\n  Actual: '
+      + show(actual)
+    );
+  }
+
+  function invalidContext(it, actual){
+    throw new TypeError(
+      it + ' was invoked outside the context of a Concurrently. \n  Called on: ' + show(actual)
+    );
+  }
+
+  function getReprType(Repr){
+    return Repr[$$type] || 'concurrify/' + (Repr.name || 'Anonymous');
+  }
+
+  function computeType(type){
+    var parts = type.split('/');
+    return parts[0] + '/Concurrent' + parts[1];
+  }
+
+  //concurrify :: Applicative m
+  //           => (TypeRep m, () -> m a, (m a, m a) -> m a, (m a, m (a -> b)) -> m b)
+  //           -> Concurrently m
+  return function concurrify(Repr, zero, race, ap){
+
+    if(!isApplicativeRepr(Repr)) invalidArgument('concurrify', 0, 'represent an Applicative', Repr);
+    if(!isFunction(zero)) invalidArgument('concurrify', 1, 'be a function', zero);
+    if(!isFunction(race)) invalidArgument('concurrify', 2, 'be a function', race);
+    if(!isBinary(race)) invalidArgument('concurrify', 2, 'be binary', race);
+    if(!isFunction(ap)) invalidArgument('concurrify', 3, 'be a function', ap);
+    if(!isBinary(ap)) invalidArgument('concurrify', 3, 'be binary', ap);
+
+    var INNERTYPE = getReprType(Repr);
+    var OUTERTYPE = computeType(INNERTYPE);
+
+    function Concurrently(sequential){
+      this.sequential = sequential;
+    }
+
+    function isConcurrently(x){
+      return x instanceof Concurrently || type(x) === OUTERTYPE;
+    }
+
+    function construct(x){
+      if(x.constructor !== Repr)
```

Comment:
`isRepr = x => x instanceof Repr || x.constructor === Repr || type(x) === Repr[$$type]`

First check gets high performance for the 99%, second check for FL compliant types constructed without a prototype chain, third for values from embedded run-times that indicate their types via sanctuary-type-identifiers.